### PR TITLE
feat(search): Shake invalid tokens

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Gives invalid tokes a little shake when they turn invalid

https://user-images.githubusercontent.com/1421724/125367854-cac6ed00-e32d-11eb-9780-875cc63066e0.mov

### CODE REVIEW

File: static/app/styles/animations.tsx [hunk 1]
```diff
@@ -124,3 +123,28 @@ export const alertHighlight = (priority: string, theme: Theme) => keyframes`
...
+    return `${i * (99 / steps)}% {
+      transform: translate(${rand1}px, ${rand2}px);
+    }`;
+  })
+  .join('\n')}
+
+  100% {
+    transform: none;
+  }
+`;
```
Comment (EvanPurkhiser) [inline static/app/styles/animations.tsx:150]: SOmetimes I feel like javascript is just getting harder and harder to read

File: static/app/styles/animations.tsx [hunk 2]
```diff
@@ -124,3 +123,28 @@ export const alertHighlight = (priority: string, theme: Theme) => keyframes`
...
+}: {steps?: number; intensity?: number | [number, number]} = {}) => keyframes`
+${new Array(steps)
+  .fill(0)
+  .map((_, i) => {
+    const [intensityX, intensityY] = Array.isArray(intensity)
+      ? intensity
+      : [intensity, intensity];
+
+    const rand1 = Math.round(Math.random() * intensityX);
+    const rand2 = Math.round(Math.random() * intensityY);
```
Comment (markstory) [inline static/app/styles/animations.tsx:139]: Should this animation be random? You could end up with subtle movement or much more in a non-deterministic way. If you had a fixed 'jiggle' amount you'd get more consistent results.

Comment (EvanPurkhiser) [reply static/app/styles/animations.tsx:139]: Yeah you're right

Comment (EvanPurkhiser) [reply static/app/styles/animations.tsx:139]: It might be tricky to get the easing right

**Review by k-fish — COMMENTED** [review]
You should maybe check `window.matchMedia("(prefers-reduced-motion: reduce)")` to disable the shake if people have that enabled.

File: static/app/styles/animations.tsx (hunk 1 above)
Comment (robinrendle) [reply static/app/styles/animations.tsx:150]: Yes

File: static/app/components/searchSyntax/renderer.tsx [hunk 3]
```diff
@@ -92,6 +106,26 @@ const FilterToken = ({
...
+      return;
+    }
+
+    const style = filterElementRef.current.style;
+
+    style.animation = 'none';
+    void filterElementRef.current.offsetTop;
+
+    window.requestAnimationFrame(
+      () => (style.animation = `${shakeAnimation.name} 300ms`)
```
Comment (EvanPurkhiser) [inline static/app/components/searchSyntax/renderer.tsx:125]: I was hoping I could let emotion take care of applying and removing the animation, but unfortunately it seems like it does it more often than I would like it to, so we have to manually handle this here

File: static/app/components/searchSyntax/renderer.tsx [hunk 4]
```diff
@@ -67,6 +68,16 @@ function renderToken(token: TokenResult<Token>, cursor: number) {
   }
 }
 
+// XXX(epurkhiser): We have to animate `left` here instead of `transform` since
+// inline elements cannot be transformed. The filter _must_ be inline to
+// support text wrapping.
+const shakeAnimation = keyframes`
+  ${new Array(4)
+    .fill(0)
+    .map((_, i) => `${i * (100 / 4)}% { left: ${3 * (i % 2 === 0 ? 1 : -1)}px; }`)
```
Comment (EvanPurkhiser) [inline static/app/components/searchSyntax/renderer.tsx:77]: procedural magic

File: static/app/styles/animations.tsx (hunk 2 above)
Comment (EvanPurkhiser) [reply static/app/styles/animations.tsx]: Ok just removed it from here and simplified a bunch

**Comment by EvanPurkhiser** [comment]
This is ready for review again now

**Review by davidenwang — APPROVED** [review]

**Review by robinrendle — APPROVED** [review]
Design looks good to me, ignoring the code though because ~stress~ feelings.

File: static/app/components/searchSyntax/renderer.tsx [hunk 5]
```diff
@@ -67,6 +68,16 @@ function renderToken(token: TokenResult<Token>, cursor: number) {
   }
 }
 
+// XXX(epurkhiser): We have to animate `left` here instead of `transform` since
```
Comment (vuluongj20) [inline static/app/components/searchSyntax/renderer.tsx:71]: You can use `display: inline-block` to get both inline wrapping behaviors and more efficient transform animations.

**Pull request merged**